---
title: Preliminary exploration of geographic trends in P. falciparum relatedness on the Colombian coast
header-includes:
output: pdf_document
---

## To-do list
- Add CIs to proportions
- Improve model fit
- Re-add sensitivity to the threshold
- How might we consolidate the two FST stories (published and this one here)
- Is there LD? 
- Incorporate MAPs here 
- Make allele frequnencies prettier

# Note
On Friday 1st of Dec, I  realised the lack of correlation between IBS and IBD was mainly due to not transposing the results of appply() and then also due to re-ordering SNPs for hmmIBD but not in the SNP data used to calculate IBS. This resulted in IBS estimates based on scrabbled numbers, which therefore didn't correlate with IBD. The IBD analyses still stand but the IBS results have changed. With a well chose cut-off, IBS performs as well as IBD in the Colombian setting. 

# Summary of FST analyses
At the city level
Using hudson's estimator, I recover insignificant p-values that do not vary in any systematic way with distance. P-values generated by permutation of city labels suggest insignificance, as do confidence intervals generated using an idenpendent method of bootstrapping over SNPs. The point estimates agree almost perfectly with estimates generated using Weir and Hill's estimator. They do not agree with estimates generated using my Weir and Cockerhams estimator however, which as expected under theoretical results reported in literature were demostrably sensitive to unbalanced sample sizes. Of those that can be compared at the state, estimates did not agree with those previously published. 

At the state level
To explore the discrepency between the previously published comparable estimate and my estimated values, I reestimated FST at the state level, and compared them with estimates generated using a independent package, DivRsity, which apparently uses Weir & Cockerham, 1984 estimator. Estimates generated using Hudson's and Wier and Hills estimator almost perfectly match estimates generated using DivRsity. My Wier and Cockerham estimates don't extactly match DivRsity, suggesting either my Weir and Cockerham estimator is wrong or DivRsity includes a sample size correction. In either case, DivRsity estimates do not agree with GENALEX6. 

# For skype with Diego: 
- Why "As expected, the lowest differentiation between the provinces was found between Valle-Nariño (FST = 0.023) while the highest was for Cauca–Chocó (FST = 0.117) (Additional file 4)."
- Confirm missing '--'
- I have treated het calls as missing data since all the samples were previously categorized as single-genotype
- Apologise for not seeing the lon/lat
- Which estimator does GENALEX 6 use? 
- How to consolidate story

### Introduction
In the following I have quantified the trend in parasite relatedness (based on identity by descent, IBD, estimated under a hidden Markov model, HMM, described elsewhere) with inter-state distance (km), with a view to assessing the potential for comparison with the Thai-Myanmar border, where, on average, the log-odds of relatedness decrease by 0.02 with every kilometer between collection sites and week between collection dates. It is a preliminary analysis, in which I have treated calls labelled '--' as missing.  

```{r setup, include=FALSE}
rm(list = ls())
knitr::opts_chunk$set(cache = TRUE, cache.comments = FALSE, echo = FALSE, include = FALSE, fig.width = 5, fig.height = 5)

library(plotrix) # For gap.barplot
library(ggplot2) # For scatter+histogram
library(ggExtra) # For scatter+histogram
library(arm) # invlogit
library(Hmisc) # cut2
library(knitr) # For kable
library(RColorBrewer)
require(gtools) # for combinations

# Load geo_dist info
load('/Users/aimeet/Documents/BroadLaptop/ColombianBarcode/RData/geo_dist_info.RData')
load('/Users/aimeet/Documents/BroadLaptop/ColombianBarcode/RData/SNPData.RData')
source('/Users/aimeet/Documents/BroadLaptop/TM_border/QuantLocalPfConnIBD/FunctionFiles/simtests.R')
attach(geo_dist_info)

unique(as.vector(as.matrix(SNPData[,6:255]))) # only 1, 0 and NA
```

```{r, include = TRUE}
# Counts of samples per city and state and time
SNPData$Year <- do.call(rbind, strsplit(as.character(SNPData$COLLECTION.DATE), split = "/"))[,3]
counts_table <- array(dim = c(length(unique(SNPData$City))+1, length(unique(SNPData$Year))+1), 
                      dimnames = list(c(unique(SNPData$City), 'Total'),
                                      c(sort(unique(SNPData$Year)), 'Total')))

for(city in rownames(counts_table)){
  for(year in colnames(counts_table)){
    counts_table[city, as.character(year)] <- sum(SNPData$City == city & SNPData$Year == year)
  }
}
counts_table[,'Total'] <- rowSums(counts_table)
counts_table['Total',] <- colSums(counts_table)
kable(counts_table)
```
```{r, MAF include = TRUE}
# MAF (almost all 0.5) 
Frequencies <- colMeans(SNPData[,6:255], na.rm = TRUE)
hist(1-Frequencies, breaks = 20)
```

```{r, include = TRUE}
#===================================================================
# Existing FST analysis (plotted against inter-state distance)
#===================================================================
# FST copy and pasted from additional file 4
# Cauca	Chocó	Nariño	Valle	 
# 0	 	 	 	             Cauca
# 0.117	0	 	 	         Chocó
# 0.096	0.106	0	 	     Nariño
# 0.073	0.076	0.023	0	 Valle

# As expected, the lowest differentiation between the provinces was found between Valle-Nariño (FST = 0.023) while the highest was for Cauca–Chocó (FST = 0.117) (Additional file 4).

# 1) Plot existing FST against distance (order names to match State_comps below)
FST_state <- c(Cauca_Choco = 0.117,
               Narino_Cauca = 0.096, 
               Narino_Choco = 0.106, 
               Valle_Cauca = 0.073, 
               Valle_Choco = 0.076, 
               Narino_Valle = 0.023)

Geo_dist_state <- rep(NA, length(FST_state))
names(Geo_dist_state) <- names(FST_state)

# Replace city names with site names  
State_comps <- names(pairwise_site_distance_all)
State_comps <- gsub('Tado', 'Choco', State_comps)
State_comps <- gsub('Quibdo', 'Choco', State_comps)
State_comps <- gsub('Buenaventura', 'Valle', State_comps)
State_comps <- gsub('Guapi', 'Cauca', State_comps)
State_comps <- gsub('Tumaco', 'Narino', State_comps)

table_state_comps <- table(State_comps)
for(comp in names(table_state_comps)){
  Geo_dist_state[comp] <- mean(pairwise_site_distance_all[comp == State_comps])
}

# Plot existing estimates
par(family = 'serif')
plot(y = FST_state, x = Geo_dist_state[names(FST_state)], 
     bty = 'n', ylab = expression('Previously published'~F[ST]), xlab = 'Inter-state distance (km)', bg = 'blue', pch = 21, 
     panel.first = grid())
text(y = FST_state, x = Geo_dist_state[names(FST_state)], 
     labels = gsub('_', ' ', names(FST_state)), pos = c(4,4,2,1,3,4))
```

```{r, include = TRUE}
#===================================================================
# New FST analysis 
#===================================================================
load('~/Documents/BroadLaptop/ColombianBarcode/RData/Fst_barcode.RData')
attach(Fst_barcode, warn.conflicts = FALSE)
comparison_names <- names(Pair_wise_site_comparisons_Fst)
par(family = 'serif')

#------------------------------------------------------------------
# Multilocus estimates
#------------------------------------------------------------------
plot(y = Pair_wise_site_comparisons_Fst, 
     x = pairwise_site_distance_all[comparison_names], 
     bty = 'n', ylab = expression(F[ST]^Hudson), xlab = 'Inter-state distance (km)', 
     bg = 'blue', pch = 21, xlim = c(50, 550),   
     panel.first = grid(), ylim = range(Pair_wise_site_comparisons_Fst_CIs))
segments(x0 = pairwise_site_distance_all[comparison_names], x1 = pairwise_site_distance_all[comparison_names],
         y0 = Pair_wise_site_comparisons_Fst_CIs[comparison_names, '2.5%'], 
         y1 = Pair_wise_site_comparisons_Fst_CIs[comparison_names, '97.5%'])
text(y = Pair_wise_site_comparisons_Fst, 
     x = pairwise_site_distance_all[comparison_names], 
     labels = gsub('_', ' ', comparison_names), 
     pos = c(4,3,4,4,4,3,4,4,4,4), cex = 0.5)
abline(h = 0, lty = 'dotted')
```

```{r}
#------------------------------------------------------------------
# SNP wise as violin plots
#------------------------------------------------------------------
plot(NULL, xlim = range(pairwise_site_distance_all[comparison_names]), 
     ylim = range(Pair_wise_site_comparisons_Fst_perSNP), panel.first = grid(),
     ylab = expression('SNP-wise'~F[ST]), xlab = 'Inter-state distance (km)', )
for(i in comparison_names){
  violin_plot(X = Pair_wise_site_comparisons_Fst_perSNP[,i], col = 'blue',
              at = pairwise_site_distance_all[i], add = TRUE, violin_width = 7)
}
points(y = Pair_wise_site_comparisons_Fst, 
       x = pairwise_site_distance_all[comparison_names], pch = 4, cex = 5)
```


```{r}
#------------------------------------------------------------------
# Significance of Fst estimates 
#------------------------------------------------------------------
# fst results store
fst_significance <- array(dim = c(length(geo_order), 1), 
                          dimnames = list(sapply(strsplit(geo_order, '_'), paste, collapse = ' '), 
                                          'pvalue'))

# Averaged over 2001-2010 (all Monte Carlo estimates)
fst_significance[,'pvalue'] <- t(apply(matrix(geo_order), 1, function(x){ 
  sim <- Pair_wise_site_comparisons_Fst_Permuted[x,]
  obs <- Pair_wise_site_comparisons_Fst[x]
  pvalue <- mcarlo_rtest(sim, obs)$pvalue_right # as in as.rtest but with n specified
  return(pvalue)
}))

kable(fst_significance) # non-significant
```


```{r, include = TRUE, fig.width=10}
#===================================================================
# FST estimates using different estimators
#===================================================================
# 1) Compare estimators 
source('/Users/aimeet/Documents/BroadLaptop/TM_border/QuantLocalPfConnIBD/FunctionFiles/calculate_pairwise_Fst.R')
Pair_wise_site_comparisons <- pairwise_site_distance[,c(1,2)]
numComparisons <- nrow(Pair_wise_site_comparisons)
SNPDataBinary <- SNPData[,6:255]
numSNPs <- ncol(SNPDataBinary)
numSamples <- nrow(SNPDataBinary)
estimators <- c('reich', 'WandC', 'WandH')
Pair_wise_site_comparisons_Fst <- array(dim = c(numComparisons,length(estimators)+1),
                                        dimnames = list(geo_order, c(estimators, 'GENALEX6')))
Sample_size_ratio <- array(dim = numComparisons,dimnames = list(geo_order))

for(i in 1:numComparisons){
  
  # Indices for populations to compare
  P1 <- (SNPData$City == as.character(Pair_wise_site_comparisons[i,1]))
  P2 <- (SNPData$City == as.character(Pair_wise_site_comparisons[i,2]))
  Sample_size_ratio[i] <- max(sum(P1)/sum(P2), sum(P2)/sum(P1)) 
  
  # Calculate Fst
  Pair_wise_site_comparisons_Fst[i,'reich'] <- calculate_pairwise_reich(P1, P2, SNPDataBinary)$F_st
  Pair_wise_site_comparisons_Fst[i,'WandC'] <- calculate_pairwise_WandC(P1, P2, SNPDataBinary)$theta_loci_hat_combined
  Pair_wise_site_comparisons_Fst[i,'WandH'] <- calculate_pairwise_WandH(P1, P2, SNPDataBinary)
}

# Add State that do not involve Choco (Tado and Quibdo) by hand
Pair_wise_site_comparisons_Fst['Guapi_Tumaco','GENALEX6'] <- FST_state['Cauca_Narino']
Pair_wise_site_comparisons_Fst['Buenaventura_Guapi','GENALEX6'] <- FST_state['Valle_Cauca']
Pair_wise_site_comparisons_Fst['Buenaventura_Tumaco','GENALEX6'] <- FST_state['Valle_Narino']

# Reich and WandH agree
par(mfrow = c(1,2), family = 'serif')
plot(x = Pair_wise_site_comparisons_Fst[,'reich'], 
     y = Pair_wise_site_comparisons_Fst[,'WandH'], 
     xlab = 'Hudson\'s estimator as outlined in Reich et al. 2009', 
     ylab = 'Weir and Hill estimator', bty = 'n', pch = 21, bg = 'gray' )
abline(a = 0, b = 1, lty = 'dotted')

# Order of magnitude difference between WandC and Reich. No agreement with previously published
# WandC is a function of sample size difference, as expected. 
plot(y = Pair_wise_site_comparisons_Fst[,'WandC'], x = Sample_size_ratio, 
     xlab = expression(n[1]/n[2]), ylab = expression(F[ST]^WC), bty = 'n', 
     pch = 21, bg = 'gray')
CorTest <- cor.test(y = Pair_wise_site_comparisons_Fst[,'WandC'], 
                    x = Sample_size_ratio)
text(y = 0.025, x = 6, labels = sprintf('r = %s (p-value = %s)', 
                                        round(CorTest$estimate, 3),
                                        round(CorTest$p.value, 3)))
```
```{r, DivRsity, fig.width=8, fig.height=5}
#============================================================================
# Estimates at the level of the state
#============================================================================
# First replace state names
load('/Users/aimeet/Documents/BroadLaptop/ColombianBarcode/RData/SNPData.RData')
SNPDataBinary <- SNPData[,6:255]
SNPData$STATE <- gsub('Nari\x96o', 'Narino', as.character(SNPData$STATE))
SNPData$STATE <- gsub('Choc\x97', 'Choco', as.character(SNPData$STATE))
SNPData$STATE <- gsub('Narino ', 'Narino', as.character(SNPData$STATE))
SNPData$STATE <- gsub('Valle ', 'Valle', as.character(SNPData$STATE))
state_comparisons <- combinations(n = length(unique(SNPData$STATE)), r = 2,unique(SNPData$STATE))
state_order <- apply(state_comparisons, 1, function(x){paste(x[1], x[2], sep = "_")})
estimators <- c('reich', 'WandC', 'WandH')
Pair_wise_state_comparisons_Fst <- array(dim = c(length(state_order),length(estimators)+2),
                                         dimnames = list(state_order, c(estimators, 'GENALEX6', 'DivRsity')))
Sample_size_ratio <- array(dim = length(state_order),dimnames = list(state_order))

for(i in 1:length(state_order)){
  
  # Indices for populations to compare
  P1 <- (SNPData$STATE == as.character(state_comparisons[i,1]))
  P2 <- (SNPData$STATE == as.character(state_comparisons[i,2]))
  Sample_size_ratio[i] <- max(sum(P1)/sum(P2), sum(P2)/sum(P1)) 
  
  # Calculate Fst
  Pair_wise_state_comparisons_Fst[i,'reich'] <- calculate_pairwise_reich(P1, P2, SNPDataBinary)$F_st
  Pair_wise_state_comparisons_Fst[i,'WandC'] <- calculate_pairwise_WandC(P1, P2, SNPDataBinary)$theta_loci_hat_combined
  Pair_wise_state_comparisons_Fst[i,'WandH'] <- calculate_pairwise_WandH(P1, P2, SNPDataBinary)
}
Pair_wise_state_comparisons_Fst['Cauca_Choco','GENALEX6'] <- FST_state['Cauca_Choco']
Pair_wise_state_comparisons_Fst['Cauca_Narino','GENALEX6'] <- FST_state['Narino_Cauca']
Pair_wise_state_comparisons_Fst['Choco_Narino','GENALEX6'] <- FST_state['Narino_Choco']
Pair_wise_state_comparisons_Fst['Cauca_Valle','GENALEX6'] <- FST_state['Valle_Cauca']
Pair_wise_state_comparisons_Fst['Choco_Valle','GENALEX6'] <- FST_state['Valle_Choco']
Pair_wise_state_comparisons_Fst['Narino_Valle','GENALEX6'] <- FST_state['Narino_Valle']


#===============================================================
# FST Using an indpendendent method (DivRsity package)
#===============================================================
library('diveRsity') # See diffCalc function example for format

Choco_ind <- SNPData$STATE == 'Choco'
Valle_ind <- SNPData$STATE == 'Valle'
Cauca_ind <- SNPData$STATE == 'Cauca'
Narino_ind <- SNPData$STATE == 'Narino'
SNPData_Genepop <- t(apply(SNPDataBinary, 1, as.character))
SNPData_Genepop[SNPData_Genepop == 1] <- '0202'
SNPData_Genepop[SNPData_Genepop == 0] <- '0101'
SNPData_Genepop[is.na(SNPData_Genepop)] <- as.character('0000')
X <- cbind(paste(rownames(SNPData_Genepop), ',', sep = ''), 
           SNPData_Genepop)
Y <- rbind(c(paste(colnames(SNPDataBinary), ',', sep = ''), ''), 
           c('POP', rep('', ncol(SNPDataBinary))),
           X[Choco_ind,], # sid98 
           c('POP', rep('', ncol(SNPDataBinary))),
           X[Valle_ind,], # sid75
           c('POP', rep('', ncol(SNPDataBinary))),
           X[Cauca_ind,], # sid50
           c('POP', rep('', ncol(SNPDataBinary))),
           X[Narino_ind,]) # sid1
write.table(Y, file = 'Colombian_divRsity_data', quote = FALSE, 
            row.names = FALSE, sep = ' ')
test_result <- diffCalc(infile = 'Colombian_divRsity_data', outfile = "Colombian_divRsity_results",
                        fst = TRUE, pairwise = TRUE, bs_locus = FALSE,
                        bs_pairwise = FALSE, boots = 100, para = TRUE)

Res <- test_result$pairwise$Fst

# Add by hand
Pair_wise_state_comparisons_Fst['Cauca_Choco','DivRsity'] <- Res['sid50,', 'sid98,']
Pair_wise_state_comparisons_Fst['Cauca_Narino','DivRsity'] <- Res['sid1,', 'sid50,']
Pair_wise_state_comparisons_Fst['Choco_Narino','DivRsity'] <- Res['sid1,', 'sid98,']
Pair_wise_state_comparisons_Fst['Cauca_Valle','DivRsity'] <- Res['sid50,', 'sid75,']
Pair_wise_state_comparisons_Fst['Choco_Valle','DivRsity'] <- Res['sid75,', 'sid98,']
Pair_wise_state_comparisons_Fst['Narino_Valle','DivRsity'] <- Res['sid1,', 'sid75,']

#===============================================================
# Plot results
#===============================================================
# Order of magnitude difference between WandC and Reich, but positive correlation.
par(mfrow = c(1,2), family = 'serif')
plot(y = Pair_wise_state_comparisons_Fst[,'reich'], 
     x = Pair_wise_state_comparisons_Fst[,'DivRsity'], 
     ylab = 'Hudson\'s estimator', 
     xlab = 'DivRsity', bty = 'n', 
     pch = 21, bg = 'gray', panel.first = grid())
abline(a = 0, b = 1, lty = 'dotted')

# # Reich and WandH agree
# plot(y = Pair_wise_state_comparisons_Fst[,'WandH'], 
#      x = Pair_wise_state_comparisons_Fst[,'DivRsity'], 
#      xlab = 'DivRsity estimate', 
#      ylab = 'Weir and Hill estimator', bty = 'n', pch = 21, bg = 'gray' )
# abline(a = 0, b = 1, lty = 'dotted')
# 
# # Order of magnitude difference between WandC and Reich, but positive correlation.
# plot(y = Pair_wise_state_comparisons_Fst[,'WandC'], 
#      x = Pair_wise_state_comparisons_Fst[,'DivRsity'], 
#      xlab = 'DivRsity', 
#      ylab = 'Weir and Cockerham', bty = 'n', 
#      pch = 21, bg = 'gray')
# CT <- cor.test(y = Pair_wise_state_comparisons_Fst[,'WandC'], 
#                x = Pair_wise_state_comparisons_Fst[,'DivRsity'])
# legend('topright', legend = sprintf('r = %s (p-value = %s)', round(CT$estimate,3), round(CT$p.value,3)), bty = 'n', cex = 0.6)

# No agreement with previously published even using W&C
plot(y = Pair_wise_state_comparisons_Fst[,'GENALEX6'], 
     x = Pair_wise_state_comparisons_Fst[,'DivRsity'], 
     xlab = 'DivRsity', ylab = 'GENALEX6', bty = 'n', 
     pch = 21, bg = 'gray', panel.first = grid())
CT <- cor.test(y = Pair_wise_state_comparisons_Fst[,'GENALEX6'], 
               x = Pair_wise_state_comparisons_Fst[,'DivRsity'])
legend('topright', legend = sprintf('r = %s (p-value = %s)', round(CT$estimate,3), round(CT$p.value,3)), bty = 'n', cex = 0.6)

# Plot New state-level estimates. Note that results for Cauca_Narino, Cauca_Valle, and Valle_Narino are as before, while others are now zero. Therefore conclusion based on city-level estimates still stand. 
plot(y = Pair_wise_state_comparisons_Fst[,'reich'], x = Geo_dist_state[rownames(Pair_wise_state_comparisons_Fst)], 
     bty = 'n', ylab = expression('Previously published'~F[ST]), 
     xlab = 'Inter-state distance (km)',
     bg = 'blue', pch = 21, 
     panel.first = grid())
text(y = Pair_wise_state_comparisons_Fst[,'reich'], 
     x = Geo_dist_state[rownames(Pair_wise_state_comparisons_Fst)], 
     labels = rownames(Pair_wise_state_comparisons_Fst))
abline(h = 0, lty = 'dotted')
```



```{r}
# Define magic numbers
threshold_IBD <- 0.5
```

```{r Raw_data_ordered, include=TRUE}
# Plot the raw data 
par(mfrow = c(1,1), family = 'serif')
cols <- brewer.pal(3,'Dark2')
image(t(as.matrix(SNPData[names(sort(rowSums(SNPData[,6:255], na.rm = T))),6:255])), 
      ylab = '', xlab = '', xaxt = 'n', yaxt = 'n', col = cols)
title(ylab = sprintf('Sample ID (%s samples)', numSamples), 
      xlab = sprintf('SNP ID (%s SNPs)', numSNPs), line = 1)
legend('topleft', bty = 'n', legend =
         c('Major', 'Minor', 'Missing'), fill = c(cols[c(1,3)], adjustcolor('white')))
```

```{r}
# Load HMM results
Result <- read.delim('/Users/aimeet/Documents/BroadLaptop/ColombianBarcode/TxtData/hmmIBD_output.hmm_fract.txt')

# Load RData
load('/Users/aimeet/Documents/BroadLaptop/ColombianBarcode/RData/SNPData.RData')
SNPData$COLLECTION.DATE <- as.Date(as.character(SNPData$COLLECTION.DATE), format = "%m/%d/%Y")

# Match up sites with cities
sites <- as.character(SNPData$City)
sites[sites == "Quibd\x97"] <- 'Quibdo'
sites[sites == "Tad\x97"] <- 'Tado'
SNPData$SITES <- sites

# Generate IBS (to check IBD approach was more-or-less okay)
# Order of as.vector(dist)
nSamples <- nrow(SNPData)
ind_j <- rep(1:nSamples, (nSamples:1)-1)
ind_i <- c()
for(i in 2:nSamples){
  ind_i <- c(ind_i, i:nSamples)
}
require(proxy)
source('/Users/aimeet/Documents/BroadLaptop/TM_border/Rscripts/Archived_for_future_ref/distance_functions.R')
IBS <- as.vector(dist(SNPData[,6:255], method = bs_distance_function))
names(IBS) <- apply(cbind(rownames(SNPData)[ind_i], rownames(SNPData)[ind_j]), 1, 
                    function(x){paste(sort(x), collapse = '', sep = '')})

# Add additional columns/covariates 
Result$sample_comp <- apply(Result[,c('sample1', 'sample2')], 1, 
                            function(x){paste(sort(x), collapse = '', sep = '')})
Result$site_comp <- apply(cbind(SNPData[as.character(Result$sample1), 'SITES'], 
                                SNPData[as.character(Result$sample2), 'SITES']), 1, 
                          function(x){paste(sort(x), sep = '', collapse = '_')})
Result$IBD <- Result$fract_sites_IBD
Result$IBS <- IBS[Result$sample_comp]
Result$geo_dist <- pairwise_site_distance_all[Result$site_comp]
Result$IBD_tail <- 1*(Result$IBD > threshold_IBD) 

# What is the corresponding cut off by percentile for IBS
IBD_cdf <- ecdf(Result$IBD)
threshold_IBS <- quantile(Result$IBS, probs = IBD_cdf(threshold_IBD))
Result$IBS_tail <- 1*(Result$IBS > threshold_IBS) 
Result$Tumaco <- Result$site_comp == "Tumaco_Tumaco"
Result$Guapi <- Result$site_comp == "Guapi_Guapi"
Result$Buenaventura <- Result$site_comp == "Buenaventura_Buenaventura"
Result$Quibdo <- Result$site_comp == "Quibdo_Quibdo"
Result$Tado <- Result$site_comp == "Tado_Tado"
Result$Within <- apply(Result[,c('Tumaco', 'Guapi', 'Buenaventura', 'Quibdo', 'Tado')], 1, any)
Result$time_dist <- abs(difftime(SNPData[as.character(Result$sample1), 'COLLECTION.DATE'], 
                                 SNPData[as.character(Result$sample2), 'COLLECTION.DATE'],
                                 units = 'weeks'))
```


### Pairwise identity by state 

The empirical density of $\hat{\pi}_{\text{IBS}}$ for all pairwise sample comparisons appears to be Guassian, but with very slightly lighter tails on the left and heavier tails on the right. The empirical density $\hat{\pi}_{\text{IBD}}$ for all pairwise sample comparisons is positively skewed with some highly related parasites. $\hat{\pi}_{\text{IBS}}$ and $\hat{\pi}_{\text{IBD}}$ do not appear to be correlated.

```{r, include = TRUE, fig.width=8}
par(family = 'serif')
hist(c(0,Result$IBS), breaks = 100, plot = TRUE, freq = FALSE, col = c(rainbow(1, start = 0.40), rainbow(101, start = 0.55, end = 0.95)), xlim = c(0,1))
abline(v = threshold_IBS, lty = 'dotted')
```



```{r IBD_hist, include = TRUE, fig.width=7}
par(family = 'serif')
X <- hist(Result$fract_sites_IBD, breaks = 100, plot = FALSE)
gap.barplot(y = X$density, gap=c(4,66), # Range to be left out
            main = "", 
            xaxt = 'n', ylab = 'Density', xlab = expression(hat(pi)[IBD]), 
            ytics = c(1,4,51,round(max(X$density)),0), las = 2,
            col = c(rainbow(1, start = 0.40), rainbow(101, start = 0.55, end = 0.95)))
abline(v = 50, lty = 'dotted')
axis(side = 1, line = 0, at = seq(0,100,25), labels = seq(0,1,0.25))
```


```{r IBD_v_IBS, include = TRUE, fig.width=7}
# create dataset with 1000 normally distributed points
df <- data.frame(x = Result$fract_sites_IBD, y = Result$IBS)
# create a ggplot2 scatterplot
p <- ggplot(df, aes(x, y)) + geom_point(colour = adjustcolor('black', alpha.f = 0.5)) + theme_classic() + labs(x = expression(hat(pi)[IBD]), y = expression(hat(pi)[IBS])) + geom_hline(yintercept = threshold_IBS, color = 'blue') + geom_vline(xintercept = threshold_IBD, color = 'blue')
# add marginal histograms
ggExtra::ggMarginal(p, type = "histogram")
plot(1:10)
```


### Plot of IBD and IBS against distance 

```{r, include = TRUE, fig.height = 8, fig.width = 7}
par(mfrow = c(2,1), family = 'serif')
# Plot of proportion > 0.5 versus distance
plot(y = Result$fract_sites_IBD, x = Result$geo_dist + rnorm(nrow(Result), 0, 5), 
     bty = 'n', pch = 21, bg = 'gray', cex = 1, 
     xlab = 'Inter-city distance (km)', ylab = expression(hat(pi)[IBD]))
abline(h = 0.5, lty = 'dotted')

plot(y = Result$IBS, x = Result$geo_dist + rnorm(nrow(Result), 0, 5), 
     bty = 'n', pch = 21, bg = 'gray', cex = 1, ylim = c(0,1),
     xlab = 'Inter-city distance (km)', ylab = expression(hat(pi)[IBD]))
abline(h = 0.5, lty = 'dotted')
```

### Plot of IBD and IBS against distance 

```{r, include = TRUE, fig.height = 8, fig.width = 7}
par(mfrow = c(2,1), family = 'serif')
plot(y = Result$IBD, x = Result$time_dist, bty = 'n', 
     xlab = 'Time difference (weeks)', ylim = c(0,1), 
     ylab = expression(hat(pi)[IBD]), main = '', pch = 20, 
     col = adjustcolor('black', alpha.f = 0.2))
plot(y = Result$IBS, x = Result$time_dist, bty = 'n', 
     xlab = 'Time difference (weeks)', ylim = c(0,1), 
     ylab = expression(hat(pi)[IBS]), main = '', pch = 20, 
     col = adjustcolor('black', alpha.f = 0.2)) 
```


```{r}
### Calculate proportions by time, simply to plot 
time_signif1 <- signif(as.numeric(Result$time_dist), 1)
nrep <- 10
proportions_time <- array(dim = c(length(unique(time_signif1)), 2), 
                          dimnames = list(sort(unique(time_signif1)), c('IBD', 'IBS')))
proportions_dist <- array(dim = c(length(unique(Result$site_comp)), 2), 
                          dimnames = list(unique(Result$site_comp), c('IBD', 'IBS')))

for(interval in time_signif1){
  ind <- time_signif1 == interval 
  proportions_time[as.character(interval), 'IBD'] <- mean(Result$IBD_tail[ind])
  proportions_time[as.character(interval), 'IBS'] <- mean(Result$IBS_tail[ind])
}
for(i in unique(Result$site_comp)){
  ind <- Result$site_comp == i 
  proportions_dist[i, 'IBD'] <- mean(Result$IBD_tail[ind])
  proportions_dist[i, 'IBS'] <- mean(Result$IBS_tail[ind])
}



```

```{r, include = TRUE, fig.height = 8, fig.width = 7}
par(mfrow = c(2,1), family = 'serif')
barplot(proportions_time[, 'IBS'], las = 2, xlab = expression(Delta~'Time (weeks)'), 
        ylab = bquote(hat(pi)[IBS]>.(round(threshold_IBS, 1))), cex.names = 0.5)
barplot(proportions_time[, 'IBD'], las = 2,  xlab = expression(Delta~'Time (weeks)'),
        ylab = bquote(hat(pi)[IBD]>.(round(threshold_IBD,1))), cex.names = 0.5)
```

```{r, include = TRUE, fig.height = 8, fig.width = 7}
par(mfrow = c(2,1), family = 'serif')
intra_inter <- sort(pairwise_site_distance_all[names(pairwise_site_distance_all)[c(1:10,21:25)]])

barplot(proportions_dist[names(intra_inter), 'IBS'], las = 2, xlab = expression(Delta~'Distance (km)'), 
        ylab = bquote(hat(pi)[IBS]>.(round(threshold_IBS, 1))), cex.names = 0.5)

barplot(proportions_dist[names(intra_inter), 'IBD'], 
        las = 2,  xlab = expression(Delta~'Distance (km)'),
        ylab = bquote(hat(pi)[IBD]>.(round(threshold_IBD,1))), cex.names = 0.5)
```


```{r, include = TRUE, fig.height = 8, fig.width = 7}
plot(y = proportions_dist[, 'IBD'], 
     x = pairwise_site_distance_all[rownames(proportions_dist)], 
     bty = 'n', pch = 21, bg = 'gray', cex = 1, 
     xlab = 'Inter-city distance (km)', ylab = expression(hat(pi)[IBD]>0.5))
text(y = proportions_dist[, 'IBD'], # pos 1,2,3,4 = b, l, 
     x = pairwise_site_distance_all[rownames(proportions_dist)], 
     labels = rownames(proportions_dist), 
     cex = 0.3, 
     pos = 3)
```


```{r}
# pi_IBD versus FST (not clear pattern)
inter_city_comp <- names(Fst_barcode$Pair_wise_site_comparisons_Fst)
plot(y = Fst_barcode$Pair_wise_site_comparisons_Fst, 
     x = proportions_dist[names(Fst_barcode$Pair_wise_site_comparisons_Fst),'IBD'])
cor.test(y = Fst_barcode$Pair_wise_site_comparisons_Fst, x = proportions_dist[names(Fst_barcode$Pair_wise_site_comparisons_Fst),'IBD']) 
```

### Regression results

```{r}
# Binned residuals (borrowed from TM boder)
binned_residuals <- function(bin_by, fitted, y, yaxis_label, main_title){
  
  # Binned residual bin_by plots 
  X <- data.frame(bin_by = bin_by)
  X$residuals <- y - fitted
  if(length(unique(bin_by)) > 8){
    X$groups <- as.numeric(cut2(bin_by, g = min(length(unique(bin_by)),30)))  
  } else {
    X$groups <- as.numeric(as.factor(as.character(bin_by)))
  }
  
  Binned_res_bin_by <- aggregate(X[, 1:2], list(X$groups), mean)
  
  #CIs
  p <- mean(y)
  two_se <- 2 * sqrt((p * (1-p))/table(X$groups))
  Binned_res_bin_by$CI_low <- + two_se
  Binned_res_bin_by$CI_hig <- - two_se
  
  # Plot
  plot(y = Binned_res_bin_by$residuals, x = Binned_res_bin_by$bin_by, 
       pch = 21, bg = 'gray', bty = 'n', xlab = yaxis_label, ylab = 'Binned residuals', 
       ylim = range(Binned_res_bin_by[,-(1:2)]), main = main_title)
  lines(y = Binned_res_bin_by$CI_low, x = Binned_res_bin_by$bin_by, lty = 'dotted')
  lines(y = Binned_res_bin_by$CI_hig, x = Binned_res_bin_by$bin_by, lty = 'dotted')
  abline(h = 0, lty = 'dotted')
}

# Define models
M1 <- '~geo_dist + time_dist'
M2 <- '~geo_dist * time_dist'
M3 <- '~ geo_dist + time_dist + Tumaco + Guapi + Buenaventura + Quibdo + Tado'
```


```{r}
# IBD tail results (all data)
GLM_IBD1 <- glm(sprintf('IBD_tail %s', M1), data = Result, family = binomial(logit), na.action = na.exclude)
GLM_IBD2 <- glm(sprintf('IBD_tail %s', M2), data = Result, family = binomial(logit), na.action = na.exclude)
GLM_IBD3 <- glm(sprintf('IBD_tail %s', M3), data = Result, family = binomial(logit), na.action = na.exclude)

print(summary(GLM_IBD1))#21531
print(summary(GLM_IBD2))#21530 and interaction term insignificant
print(summary(GLM_IBD3))#20849 and all signif except Tado
# Conclusion: drop M2 from this point

binned_residuals(bin_by = GLM_IBD1$fitted.values, 
                 fitted = GLM_IBD1$fitted.values,
                 y = Result$IBD_tail, yaxis_label = 'Fitted', main_title = '')
binned_residuals(bin_by = GLM_IBD3$fitted.values, 
                 fitted = GLM_IBD3$fitted.values,
                 y = Result$IBD_tail, yaxis_label = 'Fitted', main_title = '')
# Conclusion from binned residuals: M3 better but still malny residuals outside expected

# What happens to M1 with we remover the within and focus on inter only (no point fitting M3 since inter)? 
GLM_IBD1_inter_only <- glm(sprintf('IBD_tail %s', M1), data = Result[Result$geo_dist > 0,], family = binomial(logit), na.action = na.exclude)
print(summary(GLM_IBD1_inter_only)) # Still significant and negative!
binned_residuals(bin_by = GLM_IBD1_inter_only$fitted.values, # Some residual trend
                 fitted = GLM_IBD1_inter_only$fitted.values,
                 y = Result$IBD_tail[Result$geo_dist > 0], yaxis_label = 'Fitted', main_title = '')
```


```{r}
# IBS tail results (all data)
GLM_IBS1 <- glm(sprintf('IBS_tail %s', M1), data = Result, family = binomial(logit), na.action = na.exclude)
GLM_IBS2 <- glm(sprintf('IBS_tail %s', M2), data = Result, family = binomial(logit), na.action = na.exclude)
GLM_IBS3 <- glm(sprintf('IBS_tail %s', M3), data = Result, family = binomial(logit), na.action = na.exclude)

print(summary(GLM_IBS1))#21531 (geo dist not significant)
print(summary(GLM_IBS2))#21530 and interaction term insignificant (slightly significant)
print(summary(GLM_IBS3))#20849 and all signif except Tado

binned_residuals(bin_by = GLM_IBS1$fitted.values, 
                 fitted = GLM_IBS1$fitted.values,
                 y = Result$IBS_tail, yaxis_label = 'Fitted', main_title = '')
binned_residuals(bin_by = GLM_IBS3$fitted.values, 
                 fitted = GLM_IBS3$fitted.values,
                 y = Result$IBS_tail, yaxis_label = 'Fitted', main_title = '')
# Conclusion from binned residuals: Fits to IBS are pretty good

# What happens to M1 with we remover the within and focus on inter only (no point fitting M3 since inter)? 
GLM_IBS1_inter_only <- glm(sprintf('IBS_tail %s', M1), data = Result[Result$geo_dist > 0,], family = binomial(logit), na.action = na.exclude)
print(summary(GLM_IBS1_inter_only)) # Good fit
binned_residuals(bin_by = GLM_IBS1_inter_only$fitted.values, 
                 fitted = GLM_IBS1_inter_only$fitted.values,
                 y = Result$IBS_tail[Result$geo_dist > 0], yaxis_label = 'Fitted', main_title = '')
```

```{r}
coefs <- names(GLM_IBS3$coefficients)
Reg_rsults <- array(dim = c(length(coefs), 6), 
                    dimnames = list(coefs, 
                                    c('M1_inter_IBD', 'M1_all_IBD', 'M3_all_IBD', 
                                      'M1_inter_IBS', 'M1_all_IBS', 'M3_all_IBS')))

# Format into a table
sig3 <- function(x){formatC(signif(x,3), format='e', digits=2)} # Format scientific
fmt3 <- function(x){formatC(round(x, 3), format='f', digits=3)} # Format trailing zeros

tabular_format <- function(x){
  y <- summary(x)$coefficients
  z <- y[,c('Estimate', 'Pr(>|z|)')] 
  apply(z, 1, function(x){sprintf('%s (%s)', sig3(x[1]), fmt3(x[2]))})
}

M1_inter_IBD <- tabular_format(GLM_IBD1_inter_only)
M1_all_IBD <- tabular_format(GLM_IBD1)
M3_all_IBD <- tabular_format(GLM_IBD3)
M1_inter_IBS <- tabular_format(GLM_IBS1_inter_only)
M1_all_IBS <- tabular_format(GLM_IBS1)
M3_all_IBS <- tabular_format(GLM_IBS3)


Reg_rsults[names(M1_all_IBS), 'M1_all_IBS'] <- M1_all_IBS
Reg_rsults[names(M3_all_IBS), 'M3_all_IBS'] <- M3_all_IBS
Reg_rsults[names(M1_all_IBD), 'M1_all_IBD'] <- M1_all_IBD
Reg_rsults[names(M3_all_IBD), 'M3_all_IBD'] <- M3_all_IBD
Reg_rsults[names(M1_inter_IBS), 'M1_inter_IBS'] <- M1_inter_IBS
Reg_rsults[names(M1_inter_IBD), 'M1_inter_IBD'] <- M1_inter_IBD

kable(Reg_rsults)



```


```{r}
# IBD tail results (all data)
LM_IBD1 <- lm(sprintf('IBD %s', M1), data = Result,  na.action = na.exclude)
LM_IBD2 <- lm(sprintf('IBD %s', M2), data = Result,  na.action = na.exclude)
LM_IBD3 <- lm(sprintf('IBD %s', M3), data = Result,  na.action = na.exclude)

print(summary(LM_IBD1))# All highly signif negative
print(summary(LM_IBD2))# All highly signif 
print(summary(LM_IBD3))# All highly signif 
# Conclusion: drop M2 from this point

# What happens to M1 with we remover the within and focus on inter only (no point fitting M3 since inter)? 
LM_IBD1_inter_only <- lm(sprintf('IBD %s', M1), data = Result[Result$geo_dist > 0,],  na.action = na.exclude)
print(summary(LM_IBD1_inter_only)) # Still significant and negative!


# IBS tail results (all data)
LM_IBS1 <- lm(sprintf('IBS %s', M1), data = Result,  na.action = na.exclude)
LM_IBS2 <- lm(sprintf('IBS %s', M2), data = Result,  na.action = na.exclude)
LM_IBS3 <- lm(sprintf('IBS %s', M3), data = Result,  na.action = na.exclude)

print(summary(LM_IBS1))# All highly signif negative
print(summary(LM_IBS2))# All highly signif 
print(summary(LM_IBS3))# All highly signif 
# Conclusion: drop M2 from this point

# What happens to M1 with we remover the within and focus on inter only (no point fitting M3 since inter)? 
LM_IBS1_inter_only <- lm(sprintf('IBS %s', M1), data = Result[Result$geo_dist > 0,],  na.action = na.exclude)
print(summary(LM_IBS1_inter_only)) # Still significant and negative!
```



